doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,435 @@
+
+Frames are passed to the packetiser on a per stream basis.
+The frames must be fully formed.  By passing a frame to this function,
+ownership is passed to the packetiser which queues the frames for later
+sending by the record layer.
+
+```c
+int ossl_quic_packetiser_buffer_frame(OSSL_QUIC_TX_PACKETISER *tx,
+                                      QUIC_STREAM *stream,
+                                      const OSSL_QUIC_FRAME *frame,
+                                      size_t frame_length);
+```
+
+### Stream Send Buffers
+
+Data from the stream send buffers is treated specially.  The packetiser knows
+how much space is left in each packet and it will request that amount of data
+from the stream send buffers.  The stream send buffers will return a
+constructed frame header and a pointer to the steam data and length.  A second
+call exists to allow the packetiser to know how much data is queued for a stream
+so that planning for the creation of multiple packets is possible.
+
+```c
+int ossl_quic_get_app_data(QUIC_STREAM *stream, size_t request,
+                           const OSSL_QUIC_FRAME **frame,
+                           const unsigned char **data,
+                           size_t *data_len);
+
+size_t ossl_quic_get_app_data_size(QUIC_STREAM *stream);
+```
+
+#### Notes
+
+* Unclear how to best free the data after sent data was acked.
+  The data will be fragments from the buffers so the stream send buffers will
+  need to remember which fragment have been sent and which are pending and
+  only free once everything is sent:
+
+```c
+int ossl_quic_free_app_data(QUIC_STREAM *stream, void *data, size_t data_len);
+```
+
+* Also need a call to tell the stream send buffers to forget about previously
+  requested app data because it needs to be retransmitted and the boundaries
+  could change.  This is similar to the free call but the data is placed/queue
+  back in the send buffers appropriately, hence it might need the stream offset as
+  well:
+
+```c
+int ossl_quic_retransmitting_app_data(QUIC_STREAM *stream, size_t offset,
+                                      void *data, size_t data_len);
+```
+
+### TLS Handshake Record Layer
+
+Uses the Record Layer API to implement the inner TLS-1.3 protocol handshake.
+It produces the QUIC crypto frames which are queued using the same mechanism
+as the [Stream Send Buffers](#stream-send-buffers) above.
+
+### Flow Controller and Statistics Collector
+
+To make decisions about what frames to coalesce, the packetiser relies
+on the flow controller to enforce stream and connection bandwidth limits
+[RFC 9000 4.1 Data Flow Control].
+
+```c
+/*
+ * Return the maximum amount of data that is permitted for the given stream.
+ * This includes both the stream limit and it's associated connection limit.
+ */
+size_t ossl_quic_stream_flow_maximum_size(QUIC_STREAM *stream);
+
+/*
+ * Information the flow controller that an amount of data has been queued for
+ * sending to a stream.
```

Comment:
Is that supposed to say "Inform the flow controller"? In which case I would expect the first parameter to be a flow controller object, and the function to be named accordingly.